### ISSUE TITLE
Extend `EntityFrameworkCoreDataProtectionExtensions.PersistKeysToDbContext` to accept both TContextService and TContextImplementation

### ISSUE THREAD

**Issue by arthuredgarov** [dotnet/aspnetcore#49329]
## Summary

Currently `EntityFrameworkCoreXmlRepository<TContext>` which is being used in `PersistKeysToDbContext<TContext>()` method can only be instantiated with passing into a DbContext implementation type as a generic parameter. This is causing problems if DbContext itself is configured with `AddDbContext<TContextService, TContextImplementation>()`. I'll explain why down below.

## Motivation and goals

Lookin at the source code of `EntityFrameworkCoreXmlRepository` we can see that it is creating DbContext instances inside with `GetRequiredService<TContext>()` method.

https://github.com/dotnet/aspnetcore/blob/6b8946e25359dae624f45707ac5612b48e41d533/src/DataProtection/EntityFrameworkCore/src/EntityFrameworkCoreXmlRepository.cs#L50

And in case DbContext was configured in DI container with `AddDbContext<TContextService, TContextImplementation>()` it will retrieve an instance of DbContext, but `DbContext.Dispose()` method would be called twice. Which leads to `ObjectDisposedException` in case you have own implementation of Dispose method and working with `DbContext.ChangeTracker` there.

I tried my own implementation of `EntityFrameworkCoreXmlRepository` with both `TContextService` and `TContextImplementation` passed in as generic parameters and instead of `GetRequiredService<TContext>() (where TContext would be TContextImplementation in my case)` I am using `GetRequiredService<TContextService>()`. After such modification no double Dispose calls were noticed.

**Comment by halter73** [comment]
> And in case DbContext was configured in DI container with `AddDbContext<TContextService, TContextImplementation>()` it will retrieve an instance of DbContext, but `DbContext.Dispose()` method would be called twice. Which leads to `ObjectDisposedException` in case you have own implementation of Dispose method and working with `DbContext.ChangeTracker` there.

I don't see why `EntityFrameworkCoreXmlRepository` disposing a scope it creates would cause an `ObjectDisposedException` for other callers. Disposing a scope does not dispose any singleton services. And any disposable scoped or transient services should not be shared with any other code. We're going to need a more complete repro to understand what's going on.

https://github.com/dotnet/aspnetcore/blob/6b8946e25359dae624f45707ac5612b48e41d533/src/DataProtection/EntityFrameworkCore/src/EntityFrameworkCoreXmlRepository.cs#L48-L61

**Comment by arthuredgarov** [comment]
@halter73, [here](https://github.com/arthuredgarov/Issue49329Example) is the repository with the app to reproduce the issue. I dove more into the details and tried to debug `DependencyInjection` library classes, and currently for met it seems that this is not specifically `DataProtection` issue, rather then directly DIs.

Its quite hard for me to determine the origin of such behavior, but while running my example you would be able to see that the same instance of DbContext is being added into the service provider's list of objects to dispose twice.

**Comment by arthuredgarov** [comment]
Don't want to bother you much, but @halter73 - are there any news? Please let me know when you'd have a chance to take a look at the example. Thx

**Comment by arthuredgarov** [comment]
@halter73 One more update.

First of all - a bit more context: if we are calling `AddDbContext<TContextImplementation>()` it eventually calls `AddDbContext<TContextImplementation, TContextImplamentation>()`. So basically the logic of registering DbContext is the same for both cases. Except one thing, I'll explain down below.

Let's assume we have the following code:
```
public class MyDbContext : DbContext, IMyDbContext
{
    //Bla bla bla
}
```

If we are registering this class with `AddDbContext<MyDbContext>()` call - we would have only one service descriptor in service collection. It looks like this:
```
ServiceDescriptor:
  Service type: MyDbContext
  Implementation type: MyDbContext
  Implementation factory: null
```
But, if we are registering this class with `AddDbContext<IMyDbContext, MyDbContext>()` call - we would have two service descriptors:
```
ServiceDescriptor:
  Service type: IMyDbContext
  Implementation type: MyDbContext
  Implementation factory: null

ServiceDescriptor:
  Service type: MyDbContext
  Implementation type: null
  Implementation factory: System.Func`2[System.IServiceProvider,MyDbContext]
```

And now we are getting closer to the issue.

Since `EntityFrameworkCoreXmlRepository` is getting the context instance through the `GetRequiredService<MyDbContext>()` call it is obvious, that the second descriptor with implementation factory is being used to create an instance of the class. So we should take a closer look at this factory method:

[EntityFrameworkServiceCollectionExtensions.cs (Line 606)](https://github.com/dotnet/efcore/blob/3ef91beed8fde0015df7b4a4a4447ec30e8b9002/src/EFCore/Extensions/EntityFrameworkServiceCollectionExtensions.cs#L606)
```
if (typeof(TContextService) != typeof(TContextImplementation))
{
    serviceCollection.TryAdd(
        new ServiceDescriptor(
            typeof(TContextImplementation),
            p => (TContextImplementation)p.GetService<TContextService>()!,
            contextLifetime));
}
```

So basically when `EntityFrameworkCoreXmlRepository` is getting the context instance with `GetRequiredService<MyDbContext>()` - it calls this factory method:

`GetRequiredService<MyDbContext>()` = `p => (MyDbContext)p.GetService<IMyDbContext>()!`

and for some reason in this case service provider's scope captures the same `DbContext` instance twice into a list of objects to dispose.

This is as much information as I could provide you with. Looking forward for your response. Thanks.

**Comment by arthuredgarov** [comment]
Created a more specific [issue](https://github.com/dotnet/runtime/issues/91179).
